File: typescript/problem_55.ts
export class URLShortener55 {
    private s2l = new Map<string, string>();
    private l2s = new Map<string, string>();
    shorten(url: string): string {
        if (this.l2s.has(url)) return this.l2s.get(url)!;
        const short = Math.random().toString(36).slice(2, 8);
        this.s2l.set(short, url); this.l2s.set(url, short);
        return short;
    }
    restore(short: string): string | null { return this.s2l.get(short) ?? null; }
}

export function problem55(): void {
    console.log("Problem 55:");
    const us = new URLShortener55();
    const s1 = us.shorten("https://example.com");
    if (!s1.length) throw new Error("assertion failed");
    if (us.restore(s1) !== "https://example.com") throw new Error("assertion failed");
    if (us.shorten("https://example.com") !== s1) throw new Error("same URL same short");
    if (us.restore("notexist") !== null) throw new Error("assertion failed");
    console.log("All Tests Passed");
}

File: lua/problem_55.lua
local M = {}

function M.new_url_shortener()
    local us = {s2l={}, l2s={}}
    function us:shorten(url)
        if self.l2s[url] then return self.l2s[url] end
        local short = string.format("%06x", math.random(0, 16777215))
        self.s2l[short] = url; self.l2s[url] = short
        return short
    end
    function us:restore(short) return self.s2l[short] end
    return us
end

function M.problem55()
    print("Problem 55:")
    local us = M.new_url_shortener()
    local s1 = us:shorten("https://example.com")
    assert(s1 ~= nil and #s1 > 0)
    assert(us:restore(s1) == "https://example.com")
    assert(us:shorten("https://example.com") == s1)
    assert(us:restore("notexist") == nil)
    print("All Tests Passed")
end

return M
